File: src/test/java/dealing/with/generalization/pullUpMethodsOrFields.java
package dealing.with.generalization;

import lombok.SneakyThrows;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

public class pullUpMethodsOrFields {
    @Nested
    public class concert_should {
        private Concert concert;

        @BeforeEach
        public void init() {
            concert = new Concert("Therapie Taxi");
        }

        @SneakyThrows
        @Test
        public void retrieve_concert_duration() {
            int concertDuration = 500;

            concert.start();
            Thread.sleep(concertDuration);
            concert.stop();

            assertThat(concert.getElapsedTime()).isGreaterThanOrEqualTo(concertDuration);
        }

        @Nested
        public class throw_an_illegal_state_exception {
            @Test
            public void when_calling_start_on_an_already_started_concert() {
                concert.start();
                assertThatExceptionOfType(IllegalStateException.class)
                        .isThrownBy(() -> concert.start());
            }
        }
    }

    @Nested
    public class theater_piece_should {
        private TheaterPiece piece;

        @BeforeEach
        public void init() {
            piece = new TheaterPiece("L'invitation", 3);
        }

        @SneakyThrows
        @Test
        public void retrieve_piece_duration() {
            int pieceDuration = 500;

            piece.start();
            Thread.sleep(pieceDuration);
            piece.stop();

            assertThat(piece.getElapsedTime()).isGreaterThanOrEqualTo(pieceDuration);
        }

        @Nested
        public class throw_an_illegal_state_exception {
            @Test
            public void when_calling_start_on_an_already_started_piece() {
                piece.start();
                assertThatExceptionOfType(IllegalStateException.class)
                        .isThrownBy(() -> piece.start());
            }
        }
    }
}

File: src/scala/src/test/scala/dealing/with/generalization/pullUpMethodsOrFields.scala
package dealing.`with`.generalization

import org.scalatest.TryValues.convertTryToSuccessOrFailure
import org.scalatest.funsuite.AnyFunSuite

class pullUpMethodsOrFields extends AnyFunSuite {
  private val concert = Concert("Therapie Taxi")
  private val piece = TheaterPiece("L'invitation", 3)

  test("retrieve concert duration") {
    val concertDuration = 500

    concert.start()
    Thread.sleep(concertDuration)
    concert.stop()

    assert(concert.getElapsedTime >= concertDuration)
  }

  test(
    "concert returns a failure when calling start on an already started concert"
  ) {
    concert.start()
    assert(
      concert
        .start()
        .failure
        .exception
        .getMessage == "Concert is already in progress"
    )
  }

  test("retrieve piece duration") {
    val pieceDuration = 200

    piece.start()
    Thread.sleep(pieceDuration)
    piece.end()

    assert(piece.calculateElapsedTime >= pieceDuration)
  }

  test(
    "theater piece returns a failure when calling start on an already started piece"
  ) {
    piece.start()
    assert(
      piece
        .start()
        .failure
        .exception
        .getMessage == "Piece is already in progress"
    )
  }
}
